### PR TITLE
Task #8

### LINKED ISSUE #8: Zadanie 8 (availableSpace_)
W klasie `Player` napisz prywatną funkcję, która obliczać będzie `availableSpace_` na podstawie aktualnej ilości towaru na statku.

### PR DESCRIPTION
**Pull request by Vicx95**

### CODE REVIEW

File: shm/inc/ship.hpp [hunk 1]
```diff
@@ -1,14 +1,17 @@
 #pragma once
-
 #include <string>
+#include "cargo.hpp"
+
+#include <vector>
```
Comment (arafalski) [inline shm/inc/ship.hpp]:
```suggestion
#include <string>
#include <vector>

#include "cargo.hpp"
```

**Review by arafalski — CHANGES_REQUESTED** [review]

**Review by arafalski — DISMISSED** [review]

File: shm/inc/player.hpp [hunk 2]
```diff
@@ -17,6 +18,8 @@ class Player {
     Cargo getCargo(size_t index) const { return ship_->getCargo(index); }
 
 private:
+    size_t calculateavailableSpace();
```
Comment (kontrybutor) [inline shm/inc/player.hpp]:
```suggestion
    size_t getAvailableSpace();
```
or `calculateAvailableSpace()`

File: shm/src/player.cpp [hunk 3]
```diff
@@ -1 +1,15 @@
...
+size_t Player::calculateavailableSpace() {
+    size_t reservedSpace = 0;
+    std::for_each(ship_->getallCargo().begin(),
+                  ship_->getallCargo().end(),
+                  [&reservedSpace](const Cargo& cargo) {
+                      reservedSpace += cargo.getAmount();
+                  });
+    size_t freeSpace = ship_->getCapacity() - reservedSpace;
+
+    return freeSpace;
```
Comment (kontrybutor) [inline shm/src/player.cpp]: What do you think about:

```suggestion
size_t reservedSpace = std::accumulate(
    ship_->getAllCargo().begin(),
    ship_->getAllCargo().end(),
    0,
    [](size_t space, const Cargo& cargo) {
        return space += cargo.getAmount();
    });
size_t availableSpace = ship_->getCapacity() - reservedSpace;
```

Sorry for formatting :)

File: shm/inc/ship.hpp [hunk 4]
```diff
@@ -25,6 +26,7 @@ class Ship {
     std::string getName() const { return name_; }
     size_t getId() const { return id_; }
     Cargo getCargo(size_t index) const;
+    std::vector<Cargo>& getallCargo() { return cargo_; }
```
Comment (piezol) [inline shm/inc/ship.hpp:29]: Do we really need to make our whole cargo vector effectively public in Ship?
Idea:
In Ship a public function, that returns the available space (use solution @kontrybutor suggested).
In Player - just call that function.

**Review by piezol — CHANGES_REQUESTED** [review]

File: shm/src/ship.cpp [hunk 5]
```diff
@@ -13,8 +14,20 @@ Ship& Ship::operator+=(size_t num) {
...
+    size_t reservedSpace = std::accumulate(cargo_.begin(),
+                                           cargo_.end(),
+                                           0, [](size_t space, const Cargo& cargo) {
+                                               return space += cargo.getAmount();
+                                           });
+
+    size_t availableSpace = capacity_ - reservedSpace;
+
+    return availableSpace;
+}
```
Comment (arafalski) [inline shm/src/ship.cpp]: New line

**Review by arafalski — APPROVED** [review]

**Review by kontrybutor — APPROVED** [review]

**Pull request merged**